### ISSUE TITLE
warmup does not seem to work in libtorch backend

### ISSUE THREAD

**Issue by wangyunxiaa** [triton-inference-server/server#3012]
**Description**

**Triton Information**
What version of Triton are you using?
r21.04.2
Are you using the Triton container or did you build it yourself?
Triton container
**To Reproduce**
Steps to reproduce the behavior.
when i inference a model,the first few inference requests are slow,so I use the warmup to warm up models
l have added the following config in config.pbtxt
`model_warmup {
  name: "warmup"
  batch_size : 1
  inputs : [
    {
      key: "input__0"
      value: {
        data_type : TYPE_FP32
        dims : 1840
        zero_data : true
      }
    },
    {
      key: "length__1"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "length.bin"
      }
    },
    {
      key: "outlength__2"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "outlength.bin"
      }
    },

    {
      key: "num_id__3"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "numid.bin"
      }
    },
    {
      key: "pre_state__4"
      value: {
        data_type : TYPE_FP32
        dims : 107520
        zero_data : true
      }
    }
  ]
}
`
It seems useless, when I send a request I get 50 times time then later request, and when i change the batch size it always slower then later request with same  batchsize, It means that I need to do warmup by hand and for every batchsize.
I have tried change the config.pbtxt with batch_size=maxbatch, and add more warm_up for any batch, but the request also have above question.
I wantto know how to do warmup effectively?
Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).
name: "e2e_online_v4"
backend: "pytorch"
max_batch_size: 8
input [
  {
    name: "input__0"
    data_type: TYPE_FP32
    dims: [ 1840 ]

  }
]
input [
  {
    name: "length__1"
    data_type: TYPE_INT32
    dims: [ 1 ]

  }
]
input [
  {
    name: "outlength__2"
    data_type: TYPE_INT32
    dims: [ 1 ]

  }
]

input [
  {
    name: "num_id__3"
    data_type: TYPE_INT32
    dims: [ 1 ]

  }
]
input [
  {
    name: "pre_state__4"
    data_type: TYPE_FP32
    dims: [ 107520 ]

  }
]
output [
  {
    name: "att_output__0"
    data_type: TYPE_FP32
    dims: [ -1 ]
  }
]
output [
  {
    name: "ctc_output__1"
    data_type: TYPE_FP32
    dims: [ -1 ]
  }
]

output [
  {
    name: "cur_state__2"
    data_type: TYPE_FP32
    dims: [ -1 ]
  }
]

dynamic_batching {
  #preferred_batch_size: [ 4, 8, 16 ]
  max_queue_delay_microseconds: 15000
}

instance_group [
    {
        count: 4
        kind: KIND_GPU
    }
]

version_policy {
  all: {}
}
model_warmup {
  name: "warmup"
  batch_size : 1
  inputs : [
    {
      key: "input__0"
      value: {
        data_type : TYPE_FP32
        dims : 1840
        zero_data : true
      }
    },
    {
      key: "length__1"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "length.bin"
      }
    },
    {
      key: "outlength__2"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "outlength.bin"
      }
    },

    {
      key: "num_id__3"
      value: {
        data_type : TYPE_INT32
        dims : 1
        zero_data : true
        #input_data_file : "numid.bin"
      }
    },
    {
      key: "pre_state__4"
      value: {
        data_type : TYPE_FP32
        dims : 107520
        zero_data : true
      }
    }
  ]
}
model_warmup {
  name: "warmup1"
  batch_size : 2
  inputs : [
    {
      key: "input__0"
      value: {
        data_type : TYPE_FP32
        dims : 1840
        zero_data : true
      }
    },
    {
      key: "length__1"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "length.bin"
      }
    },
    {
      key: "outlength__2"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "outlength.bin"
      }
    },

    {
      key: "num_id__3"
      value: {
        data_type : TYPE_INT32
        dims : 1
        input_data_file : "numid.bin"
      }
    },
    {
      key: "pre_state__4"
      value: {
        data_type : TYPE_FP32
        dims : 107520
        zero_data : true
      }
    }
  ]
}
....

**Expected behavior**

**Comment by CoderHam** [comment]
@wangyunxiaa this issue was root caused to thread specific model caching. It is an issue caused with multiple pytorch model instances. Can you repeat your experiments with a single pytorch model instance?

**Comment by wangyunxiaa** [comment]
I have change the count to 1, and just run one model with one version. I write a model.py to see the batch infermation on server,
this is my test log:
`[yx.wang@dev-istio-lasr-001 triton_server]$ sudo sh run_triton_server.sh 
AISPEECH_SERVER_BASEPATH:     triton_server
AISPEECH_MODEL_PATH:         triton_models
I0616 02:06:30.441025 18 metrics.cc:221] Collecting metrics for GPU 0: GeForce GTX 1080 Ti
I0616 02:06:30.899462 18 libtorch.cc:940] TRITONBACKEND_Initialize: pytorch
I0616 02:06:30.899508 18 libtorch.cc:950] Triton TRITONBACKEND API version: 1.0
I0616 02:06:30.899519 18 libtorch.cc:956] 'pytorch' TRITONBACKEND API version: 1.0
2021-06-16 10:06:31.163283: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
I0616 02:06:31.203798 18 tensorflow.cc:1876] TRITONBACKEND_Initialize: tensorflow
I0616 02:06:31.203840 18 tensorflow.cc:1889] Triton TRITONBACKEND API version: 1.0
I0616 02:06:31.203852 18 tensorflow.cc:1895] 'tensorflow' TRITONBACKEND API version: 1.0
I0616 02:06:31.203862 18 tensorflow.cc:1916] backend configuration:
{}
I0616 02:06:31.430791 18 pinned_memory_manager.cc:205] Pinned memory pool is created at '0x7ff064000000' with size 268435456
I0616 02:06:31.431254 18 cuda_memory_manager.cc:103] CUDA memory pool is created on device 0 with size 67108864
W0616 02:06:31.433919 18 model_repository_manager.cc:2020] ignore version directory 'torchscript_model' which fails to convert to integral number
I0616 02:06:31.434033 18 model_repository_manager.cc:787] loading: e2e_online_v4:1
I0616 02:06:31.563657 18 python.cc:615] TRITONBACKEND_ModelInstanceInitialize: e2e_online_v4_0 (GPU device 0)
[W Context.cpp:69] Warning: torch.set_deterministic is in beta, and its design and  functionality may change in the future. (function operator())
----> model 1 batch size 1
infere-time-use:  1.3092105388641357
----> model 1 batch size 2
infere-time-use:  3.1362571716308594
----> model 1 batch size 3
infere-time-use:  3.074922800064087
----> model 1 batch size 4
infere-time-use:  3.214728355407715
----> model 1 batch size 5
infere-time-use:  3.1138999462127686
----> model 1 batch size 6
infere-time-use:  3.1955184936523438
----> model 1 batch size 7
infere-time-use:  3.1302871704101562
----> model 1 batch size 8
infere-time-use:  3.1774158477783203
----> model 1 batch size 1
infere-time-use:  0.011354684829711914
----> model 1 batch size 2
infere-time-use:  0.007663726806640625
----> model 1 batch size 3
infere-time-use:  0.007612466812133789
----> model 1 batch size 4
infere-time-use:  0.007318973541259766
----> model 1 batch size 5
infere-time-use:  0.007546901702880859
----> model 1 batch size 6
infere-time-use:  0.009435415267944336
----> model 1 batch size 7
infere-time-use:  0.007947444915771484
----> model 1 batch size 8
infere-time-use:  3.147199869155884
I0616 02:07:08.953415 18 model_repository_manager.cc:960] successfully loaded 'e2e_online_v4' version 1
I0616 02:07:08.954125 18 server.cc:495] 
+------------+-----------------------------------------------------------------+--------------------------------------------------------+
| Backend    | Config                                                          | Path                                                   |
+------------+-----------------------------------------------------------------+--------------------------------------------------------+
| pytorch    | /opt/tritonserver/backends/pytorch/libtriton_pytorch.so         | {}                                                     |
| tensorflow | /opt/tritonserver/backends/tensorflow1/libtriton_tensorflow1.so | {}                                                     |
| python     | /opt/tritonserver/backends/python/libtriton_python.so           | {"cmdline":{"python-runtime":"/opt/conda/bin/python"}} |
+------------+-----------------------------------------------------------------+--------------------------------------------------------+

I0616 02:07:08.954614 18 server.cc:538] 
+---------------+---------+--------+
| Model         | Version | Status |
+---------------+---------+--------+
| e2e_online_v4 | 1       | READY  |
+---------------+---------+--------+

I0616 02:07:08.955058 18 tritonserver.cc:1642] 
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------+
| Option                           | Value                                                                                                                                              |
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------+
| server_id                        | triton                                                                                                                                             |
| server_version                   | 2.7.0                                                                                                                                              |
| server_extensions                | classification sequence model_repository schedule_policy model_configuration system_shared_memory cuda_shared_memory binary_tensor_data statistics |
| model_repository_path[0]         | /models/triton_models                                                                                                                              |
| model_control_mode               | MODE_POLL                                                                                                                                          |
| strict_model_config              | 1                                                                                                                                                  |
| pinned_memory_pool_byte_size     | 268435456                                                                                                                                          |
| cuda_memory_pool_byte_size{0}    | 67108864                                                                                                                                           |
| min_supported_compute_capability | 6.0                                                                                                                                                |
| strict_readiness                 | 1                                                                                                                                                  |
| exit_timeout                     | 5                                                                                                                                                  |
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------+

I0616 02:07:08.961327 18 grpc_server.cc:3989] Started GRPCInferenceService at 0.0.0.0:8001
I0616 02:07:08.961379 18 grpc_server.cc:3991] GRPCInferenceService listen at unix:/models/f
I0616 02:07:08.962101 18 http_server.cc:2717] Started HTTPService at 0.0.0.0:8000
I0616 02:07:09.005289 18 http_server.cc:2736] Started Metrics Service at 0.0.0.0:8002
----> model 1 batch size 1
infere-time-use:  0.014166116714477539
----> model 1 batch size 1
infere-time-use:  0.012605428695678711
----> model 1 batch size 1
infere-time-use:  0.01188039779663086
----> model 1 batch size 1
infere-time-use:  0.011366605758666992
----> model 1 batch size 1
infere-time-use:  0.011696577072143555
----> model 1 batch size 1
infere-time-use:  0.010614871978759766
----> model 1 batch size 1
infere-time-use:  0.00910806655883789
----> model 1 batch size 1
infere-time-use:  0.011645317077636719
----> model 1 batch size 1
infere-time-use:  0.010998249053955078
----> model 1 batch size 1
infere-time-use:  0.008876323699951172
----> model 1 batch size 1
infere-time-use:  0.007399082183837891
----> model 1 batch size 2
infere-time-use:  0.01717519760131836
----> model 1 batch size 2
infere-time-use:  2.8507518768310547
----> model 1 batch size 2
infere-time-use:  0.010759353637695312
----> model 1 batch size 2
infere-time-use:  0.010294437408447266
----> model 1 batch size 2
infere-time-use:  0.008519172668457031
----> model 1 batch size 2
infere-time-use:  0.009655952453613281
----> model 1 batch size 2
infere-time-use:  0.009712696075439453
----> model 1 batch size 2
infere-time-use:  0.012486457824707031
----> model 1 batch size 2
infere-time-use:  0.012797355651855469
----> model 1 batch size 2
infere-time-use:  0.01392984390258789
----> model 1 batch size 2
infere-time-use:  0.011692523956298828
----> model 1 batch size 3
infere-time-use:  0.013164520263671875
----> model 1 batch size 3
infere-time-use:  2.8781588077545166
----> model 1 batch size 3
infere-time-use:  0.011850595474243164
----> model 1 batch size 3
infere-time-use:  0.012597084045410156
----> model 1 batch size 3
infere-time-use:  0.009187459945678711
----> model 1 batch size 3
infere-time-use:  0.008015155792236328
----> model 1 batch size 3
infere-time-use:  0.012218236923217773
----> model 1 batch size 3
infere-time-use:  0.011744260787963867
----> model 1 batch size 3
infere-time-use:  0.013356447219848633
----> model 1 batch size 3
infere-time-use:  0.012698650360107422
----> model 1 batch size 3
infere-time-use:  0.010189294815063477
----> model 1 batch size 1
infere-time-use:  0.012172937393188477
----> model 1 batch size 3
infere-time-use:  0.006596803665161133
----> model 1 batch size 1
infere-time-use:  0.009293079376220703
----> model 1 batch size 3
infere-time-use:  0.006487607955932617
----> model 1 batch size 1
infere-time-use:  0.0058820247650146484
----> model 1 batch size 4
infere-time-use:  0.00829315185546875
----> model 1 batch size 4
infere-time-use:  2.8436362743377686
----> model 1 batch size 4
infere-time-use:  0.012121438980102539
----> model 1 batch size 4
infere-time-use:  0.011739253997802734
----> model 1 batch size 4
infere-time-use:  0.009254693984985352
----> model 1 batch size 4
infere-time-use:  0.01253819465637207
----> model 1 batch size 4
infere-time-use:  0.011883020401000977
----> model 1 batch size 4
infere-time-use:  0.012014150619506836
----> model 1 batch size 3
infere-time-use:  0.012490510940551758
`
I have warmup the model with batchsize 1-8 twice, and when i send requests, the server do quickly when do one batch, but it is slowly sometimes when batchsize is growing

**Comment by CoderHam** [comment]
Can you repeat these experiments for warmup using the pytorch python API and the traced model? This would help us zero down on whether the issue is specific to the C API implementation.
In some cases optimized execution can cause the model to be slower and have inconsistent warmup. We recently added a flag for disabling optimized execution in pytorch which might be useful in this case.

**Comment by wangyunxiaa** [comment]
![image](https://user-images.githubusercontent.com/41035013/122367939-79764880-cf8f-11eb-9e19-97c48eafbfc8.png)
python API with torchscript model also has this issue,
python API with pytorch model no, but when trace the model add `torch.jit.optimized_execution(False)`  and `with torch.jit.optimized_execution(args.optimize)` heve no change of the result. 

maybe this is a bug of pytorch:
https://github.com/pytorch/pytorch/issues/57894
I have tried https://github.com/triton-inference-server/pytorch_backend
parameters: {
key: "DISABLE_OPTIMIZED_EXECUTION"
    value: {
    string_value:"true"
    }
}
these all above seem to be no effect

**Comment by CoderHam** [comment]
@wangyunxiaa the DISABLE_OPTIMIZED_EXECUTION feature was added recently and will be available in the 21.06 release. From your update it confirms that it is a pytorch issue. I would follow up with the pytorch folks. Closing this ticket as it is not a Triton issue.

**Comment by hyperfraise** [comment]
Did you ever manage to solve this ? My model triton speed is much lower than a simple pythonic loop.
